working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+# Shared Memory Multithreading for Dart
+
+This proposal tries to address two long standing gaps which separate Dart from other more low-level languages:
+
+* Dart developers should be able to utilize multicore capabilities without hitting limitations of the isolate model.
+
+* Interoperability between Dart and native platforms (C/C++, Objective-C/Swift, Java/Kotlin) requires aligning concurrency models between Dart world and native world. The misalignment is not an issue when invoking simple native synchronous APIs from Dart, but it becomes a blocker when:
+
+  * working with APIs pinned to a specific thread (e.g. `@MainActor` APIs in Swift or UI thread APIs on Android) 
+
+  * dealing with APIs which want to call _back_ from native into Dart on an arbitrary thread.
+
+  * using Dart as a container for shared business logic - which can be invoked on an arbitrary thread by surrounding native application. 
+
+  Native code does not _understand_ the concept of isolates. 
+
+The biggest challenge with introducing shared memory multithreading into Dart is avoiding breaking existing code, which is written under the assumption of single-threaded access. We would like to avoid just outright breaking existing code in subtle ways. Consider for a moment a library which is maintains a global cache internally:  
+
+```     dart
+library foo;
+
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId;
+final _cache = <int, C>{};
+
+C makeObject() {
+  map[nextId] = C(nextId);
+  nextId++;
+  return map[nextId];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the same code becomes thread _unsafe_ in the presence of shared memory multithreading.
+
+## Shareable Data
+
+We propose to extend Dart with a shared memory multithreading but provide a clear type-based delineation between two concurrency worlds. **Type of an instance should provide a clear indication whether its fields can be concurrently mutated by another thread.** We are going to call such types _shareable_ and assume that the language provides some way to opt-in into this for classes:
+
+```  dart
+/// Mark class sharable by implementing a [Shareable]
+/// interface.
+class S implements Shareable {
+  // ...
+} 
+```
+
+To maintain heap isolation we restrict the kind of data that a shareable class can contain: **Fields of sharable classes can only contain references to instances of sharable classes.** There is however no requirement of immutability imposed on shareable classes.
+
+References to shareable instances can be passed between isolates within an isolate group directly: for example sending a shareable instance through `SendPort` will not copy it.
+
+```dart  
+class S implements Shareable {
+  int v = 0;
+}
+
+final s = S();
```

Comment:
Am I correct in assuming we don't need a `@shared` here, because we _only_ mark existing value and immutable types (`int`, `double`, enums, `Pointer`s) as `Sharable`?

If we were to mark any existing class that has a mutable field as sharable, we would change the semantics of this example.

Reply:
it's a local variable which is captured, not a global variable. Made it more clear by wrapping it into a function.